### PR TITLE
Close cast soundness loophole

### PR DESCRIPTION
**Pull request by AlexKnauth**
Fixes https://github.com/racket/typed-racket/issues/108
Fixes PR 13626
Fixes PR 14955

This makes `cast` generate two contracts, one from typed to untyped using the original type of the expression, and one from untyped to typed, using the type it's casting to. 

So in examples like this:
```racket
#lang typed/racket

(: v Boolean)
(define v #f)
(: f (Boolean -> Boolean))
(define (f x)
  (set! v x)
  x)

v ; #f
((cast f (String -> String)) "hello") ; error!
v ; if that hadn't been an error, this would be "hello" with type Boolean
```
The cast applies a contract for the new type `String -> String` going from untyped to typed, but first it applies a contract for the old type, `Boolean -> Boolean`, going from typed to untyped. This also means that doing `((cast f (Boolean -> Boolean)) #t)` succeeds and sets `v` to `#t`, because it checked that the argument was a boolean first.

### CODE REVIEW

**Comment by samth** [comment]
Wooo!

File: typed-racket-lib/typed-racket/base-env/prims-contract.rkt [hunk 1]
```diff
@@ -308,8 +316,15 @@
      (cond [(not (unbox typed-context?)) ; no-check, don't check
             #'v]
            [else
-            (define ctc (syntax-local-lift-expression
-                         (make-contract-def-rhs #'ty #f #f)))
+            (define existing-ty-sym (gensym))
+            (define new-ty-ctc (syntax-local-lift-expression
+                                (make-contract-def-rhs #'ty #f #f)))
+            (define existing-ty-ctc (syntax-local-lift-expression
+                                     ;; TODO: This should be changed to somehow use the
```
Comment (takikawa) [inline typed-racket-lib/typed-racket/base-env/prims-contract.rkt]: This TODO comment is now unneeded right?

**Comment by takikawa** [comment]
Thanks! Can you add some test cases, preferably including the ones from those issues/PRs? Otherwise looks good to me.

**Comment by takikawa** [comment]
Oh, one more thing: we should make sure this doesn't break too many programs before shipping it in a release. For example by trying it on TR-written packages on the package catalog. If it does break a bunch of things, we should at least mention the backwards incompatibility on the mailing list.

File: typed-racket-lib/typed-racket/base-env/prims-contract.rkt [hunk 2]
```diff
@@ -250,10 +251,17 @@
...
 ;;  cast
 
-;; Helper to construct syntax for contract definitions
+;; Helpers to construct syntax for contract definitions
+;; make-contract-def-rhs : Type-Stx Boolean Boolean -> Syntax
 (define (make-contract-def-rhs type flat? maker?)
   (contract-def-property #'#f `#s(contract-def ,type ,flat? ,maker? untyped)))
 
+;; make-contract-def-rhs/from-typed : Symbol Boolean Boolean -> Syntax
+(define (make-contract-def-rhs/from-typed sym flat? maker?)
```
Comment (samth) [inline typed-racket-lib/typed-racket/base-env/prims-contract.rkt]: Why not make this a version of `make-contract-def-rhs`?

Comment (samth) [inline typed-racket-lib/typed-racket/base-env/prims-contract.rkt]: Also, can we just delay all `contract-def-property` constructions so that we have a uniform interface?

File: typed-racket-lib/typed-racket/base-env/prims-contract.rkt (hunk 1 above)
Comment (AlexKnauth) [inline typed-racket-lib/typed-racket/base-env/prims-contract.rkt]: Thanks, fixed now.

**Comment by AlexKnauth** [comment]
I've added the examples from PR 13626 and issue #108 as test cases.

File: typed-racket-lib/typed-racket/base-env/prims-contract.rkt [hunk 3]
```diff
@@ -319,7 +332,12 @@
                  "Type ~a could not be converted to a contract because it contains free variables."
                  type)))
             #`(#,(external-check-property #'#%expression check-valid-type)
-               #,(apply-contract ctc))])]))
+               #,(apply-contract
+                  (apply-contract
+                   #`(#,(casted-expr-property #'#%expression store-existing-type)
```
Comment (samth) [inline typed-racket-lib/typed-racket/base-env/prims-contract.rkt]: Why put this on the `#%expression` identifier, instead of on the whole form?

File: typed-racket-lib/typed-racket/base-env/prims-contract.rkt [hunk 4]
```diff
@@ -308,8 +316,13 @@
      (cond [(not (unbox typed-context?)) ; no-check, don't check
             #'v]
            [else
-            (define ctc (syntax-local-lift-expression
-                         (make-contract-def-rhs #'ty #f #f)))
+            (define existing-ty-sym (gensym))
+            (define new-ty-ctc (syntax-local-lift-expression
```
Comment (samth) [inline typed-racket-lib/typed-racket/base-env/prims-contract.rkt]: Could we use an id table based on `new-ty-ctc` (or something else) instead of using a gensym?

File: typed-racket-lib/typed-racket/typecheck/tc-expression.rkt [hunk 5]
```diff
@@ -34,6 +34,14 @@
...
         (tc-expr/check #'e expected)
         (tc-expr #'e))]
+    [(exp:casted-expr^ e)
+     (define result (tc-expr #'e))
+     (match result
+       [(tc-result1: ty)
+        ((attribute exp.value) ty)
+        result]
+       [_
+        (tc-error/expr "Cannot cast expression that produces multiple values")])]
```
Comment (samth) [inline typed-racket-lib/typed-racket/typecheck/tc-expression.rkt]: Is there a reason we can't support this?

File: typed-racket-lib/typed-racket/base-env/prims-contract.rkt (hunk 4 above)
Comment (AlexKnauth) [inline typed-racket-lib/typed-racket/base-env/prims-contract.rkt]: I could make it an id-table that used `(generate-temporary)`, would that be better?

File: typed-racket-lib/typed-racket/base-env/prims-contract.rkt [hunk 6]
```diff
@@ -254,11 +254,12 @@
 ;; Helpers to construct syntax for contract definitions
 ;; make-contract-def-rhs : Type-Stx Boolean Boolean -> Syntax
 (define (make-contract-def-rhs type flat? maker?)
-  (contract-def-property #'#f `#s(contract-def ,type ,flat? ,maker? untyped)))
+  (define contract-def `#s(contract-def ,type ,flat? ,maker? untyped))
+  (contract-def-property #'#f (λ () contract-def)))
```
Comment (samth) [inline typed-racket-lib/typed-racket/base-env/prims-contract.rkt]: Let's use promises instead of thunks.

File: typed-racket-lib/typed-racket/base-env/prims-contract.rkt (hunk 3 above)
Comment (AlexKnauth) [inline typed-racket-lib/typed-racket/base-env/prims-contract.rkt]: Because that's where the `tc/#%expression` function looks for it, and that's what it already does for the `external-check-property` above it.

File: typed-racket-lib/typed-racket/base-env/prims-contract.rkt (hunk 4 above)
Comment (samth) [inline typed-racket-lib/typed-racket/base-env/prims-contract.rkt]: Won't the identifier that you get from `syntax-local-lift-expression` be unique?

File: typed-racket-lib/typed-racket/base-env/prims-contract.rkt (hunk 6 above)
Comment (AlexKnauth) [inline typed-racket-lib/typed-racket/base-env/prims-contract.rkt]: I would be little bit uncomfortable with that, because even though promises delay computation, it's usually for a different reason than this. Usually when you see a promise, a forcing it works in the same way that evaluating it in the original context did, and mutation isn't supposed to change that. In this case, we're delaying the computation specifically so that it can rely on a mutation that will happen later, so promises feel a bit wrong to me.

File: typed-racket-lib/typed-racket/typecheck/tc-expression.rkt (hunk 5 above)
Comment (AlexKnauth) [inline typed-racket-lib/typed-racket/typecheck/tc-expression.rkt]: The [`contract`](http://docs.racket-lang.org/reference/attaching-contracts-to-values.html#%28form._%28%28lib._racket%2Fcontract%2Fprivate%2Fbase..rkt%29._contract%29%29) form only supports single values, right?

Comment (samth) [inline typed-racket-lib/typed-racket/typecheck/tc-expression.rkt]: We could implement something with `call-with-values`.

Comment (AlexKnauth) [inline typed-racket-lib/typed-racket/typecheck/tc-expression.rkt]: Yes. But that's complicated enough that it should be a separate pull request, right?

Comment (samth) [inline typed-racket-lib/typed-racket/typecheck/tc-expression.rkt]: Sure.

File: typed-racket-lib/typed-racket/base-env/prims-contract.rkt [hunk 7]
```diff
@@ -250,9 +251,17 @@
 ;;  make-predicate
 ;;  cast
 
-;; Helper to construct syntax for contract definitions
+;; Helpers to construct syntax for contract definitions
+;; make-contract-def-rhs : Type-Stx Boolean Boolean -> Syntax
 (define (make-contract-def-rhs type flat? maker?)
-  (contract-def-property #'#f `#s(contract-def ,type ,flat? ,maker? untyped)))
+  (define contract-def `#s(contract-def ,type ,flat? ,maker? untyped))
+  (contract-def-property #'#f (λ () contract-def)))
```
Comment (AlexKnauth) [inline typed-racket-lib/typed-racket/base-env/prims-contract.rkt]: From https://github.com/racket/typed-racket/pull/365#discussion_r64797153 which is an "outdated diff" for some reason:

@samth :
Let's use promises instead of thunks.

@AlexKnauth :
I would be little bit uncomfortable with that, because even though promises delay computation, it's usually for a different reason than this. Usually when you see a promise, a forcing it works in the same way that evaluating it in the original context would, and mutation isn't "supposed to" change that. In this case, we're delaying the computation specifically so that it can rely on a mutation that will happen later, so promises feel a bit wrong to me.

Does this make sense?

Comment (AlexKnauth) [inline typed-racket-lib/typed-racket/base-env/prims-contract.rkt]: I just thought of something else. If it were a promise, than an "accidental" too-early force would cache the wrong result, and the contract-generation pass would generate a wrapping `Any` contract even though by that time the `tc/#%expression` has put the correct type in the table. Using lambda means it might recompute a few more things, but that gets rid of the danger that a too-early application will ruin the result for future function applications of the lambda.

Comment (samth) [inline typed-racket-lib/typed-racket/base-env/prims-contract.rkt]: Can we make a too-early force or call an error?

Comment (AlexKnauth) [inline typed-racket-lib/typed-racket/base-env/prims-contract.rkt]: No, because if type-checking somehow stops in a way that it never sees the `casted-expr` property, then it will seem like a too-early force, and in that case it makes sense to generate the wrapping `Any` contract as a fallback.

Comment (samth) [inline typed-racket-lib/typed-racket/base-env/prims-contract.rkt]: Why would typechecking stop, but this still get forced?

Comment (AlexKnauth) [inline typed-racket-lib/typed-racket/base-env/prims-contract.rkt]: I don't know, it might be impossible. I don't know very much about how type errors get reported here, how the contract generation phase works, or how the type-checker checks typed terms within `ignore-some` expressions, so I'm kind of paranoid. But still, lambda seems safer, even if promises are also safe in this case, the way the rest of the code is designed right now.

Comment (samth) [inline typed-racket-lib/typed-racket/base-env/prims-contract.rkt]: Just a quick overview of how this works:

* When you report a type error and we're reporting multiple type errors, it gets stored in a list in a mutable variable.
* Then, after `pass2` in the type checker, we call `report-all-errors`, which will raise a syntax error if needed. That code is [here](https://github.com/racket/typed-racket/blob/master/typed-racket-lib/typed-racket/typecheck/tc-toplevel.rkt#L409). 
* The contract generation happens after this, when we (generate the provides)[https://github.com/racket/typed-racket/blob/master/typed-racket-lib/typed-racket/typecheck/tc-toplevel.rkt#L445]. 

I'm ok with sticking with the `lambda`, but I'd like to make this an error if we call the thunk to early.

File: typed-racket-lib/typed-racket/base-env/prims-contract.rkt [hunk 8]
```diff
@@ -250,9 +251,17 @@
...
+
+;; make-contract-def-rhs/from-typed : Id Boolean Boolean -> Syntax
+(define (make-contract-def-rhs/from-typed id flat? maker?)
+  (contract-def-property
+   #'#f
+   (λ () `#s(contract-def ,(or (cast-table-ref id) #'Any) ,flat? ,maker? typed))))
```
Comment (AlexKnauth) [inline typed-racket-lib/typed-racket/base-env/prims-contract.rkt]: Re: https://github.com/racket/typed-racket/pull/365/files/357e4a12419fbc6af056a77710823bfc53b1a75e#r64914939

> I'd like to make this an error if we call the thunk too early

Does that mean making this `(or (cast-table-ref id) (error ...))`?

Comment (samth) [inline typed-racket-lib/typed-racket/base-env/prims-contract.rkt]: Yes.

Comment (AlexKnauth) [inline typed-racket-lib/typed-racket/base-env/prims-contract.rkt]: Hm. I tried that, and it passes the cast tests, both new and old, in a module, in the repl, and at the top level, but I'm getting that error when `special-forms.scrbl` is run, specifically because of these examples here: https://github.com/AlexKnauth/typed-racket/blob/close-cast-loophole/typed-racket-doc/typed-racket/scribblings/reference/special-forms.scrbl#L542

Comment (samth) [inline typed-racket-lib/typed-racket/base-env/prims-contract.rkt]: Do those particular examples work at the repl? What about when just calling `eval` on them (in an appropriate namespace)?

Comment (AlexKnauth) [inline typed-racket-lib/typed-racket/base-env/prims-contract.rkt]: Yes, they work at the repl, using `eval`, and using `(make-base-eval)` from `scribble/example` (both with the error and with the `Any` fallback).

Comment (AlexKnauth) [inline typed-racket-lib/typed-racket/base-env/prims-contract.rkt]: Actually no, (maybe because I hadn't compiled it properly yet), now this program gives me the "thunk called too early" error:
```racket
#lang racket
(require scribble/example)
(define ev (make-base-eval))
(ev '(require typed/racket))
(ev '(cast 3 Integer))
```
Although it still works both at the repl and using `eval`.

Comment (AlexKnauth) [inline typed-racket-lib/typed-racket/base-env/prims-contract.rkt]: It might be because of the sandboxed evaluators, because this also fails in the "thunk called too early" version:
```racket
#lang racket
(require racket/sandbox)
(define ev (call-with-trusted-sandbox-configuration
             (λ () (make-module-evaluator "#lang typed/racket"))))
(ev '(cast 3 Integer))
```

Comment (AlexKnauth) [inline typed-racket-lib/typed-racket/base-env/prims-contract.rkt]: And in the `Any` fallback version, it's still generating the correct contract, so the cast-table mutation is happening correctly, and the contract-generation pass gets the right contract. So it must be something else that's calling the thunk. So in the `Any` fallback version, this:
```racket
#lang racket
(require racket/sandbox)
(define ev (call-with-trusted-sandbox-configuration
             (λ () (make-module-evaluator "#lang typed/racket"))))
(ev '((cast (lambda ([s : String]) s) (Any -> Any)) 5))
```
Generates the correct error:
```
#<procedure:val>: contract violation
  expected: String
  given: 5
  in: the 1st argument of
      (-> String any)
  contract from: typed-world
  blaming: cast
   (assuming the contract is correct)
  at: eval:6.0
```
Where if it were getting the `Any`, it would raise a different error saying `Attempted to use a higher-order value passed as `Any` in untyped code: #<procedure:val>`. Since it's not that, we know that has gotten the correct type by the time the contract-generation pass happens.

Comment (samth) [inline typed-racket-lib/typed-racket/base-env/prims-contract.rkt]: Ok, we should figure out where that error is coming from. What's the stack trace when you get the error?

Comment (AlexKnauth) [inline typed-racket-lib/typed-racket/base-env/prims-contract.rkt]:
```
Internal Typechecker Error: contract-def-property: thunk called too early
  This should only be called after the type-checking pass has finished.
while typechecking:
here
originally:
here
  context...:
   /Users/Alex/Documents/GitHub/typed-racket/typed-racket-lib/typed-racket/utils/tc-utils.rkt:270:0: int-err
   /Users/Alex/Documents/GitHub/typed-racket/typed-racket-lib/typed-racket/base-env/prims-contract.rkt:268:3
   /Users/Alex/Documents/GitHub/typed-racket/typed-racket-lib/typed-racket/private/type-contract.rkt:82:0: generate-contract-def
   /Users/Alex/Documents/GitHub/typed-racket/typed-racket-lib/typed-racket/private/type-contract.rkt:195:0: change-contract-fixups
   /Users/Alex/Documents/GitHub/typed-racket/typed-racket-lib/typed-racket/typecheck/toplevel-trampoline.rkt:108:24: for-loop
   dots-loop
   /applications/racket/2016-05-23/tr/Racket v6.5.0.5/collects/racket/private/more-scheme.rkt:148:2: call-with-break-parameterization
   .../more-scheme.rkt:261:28
   /applications/racket/2016-05-23/tr/Racket v6.5.0.5/share/pkgs/sandbox-lib/racket/sandbox.rkt:843:5: loop
```
It seems like it's going though `generate-contract-def` and `change-contract-fixups`, which are supposed to be for the contract-generation phase, I think. Is there any chance that the contract-generation phase is happening before type-checking is finished?

Comment (samth) [inline typed-racket-lib/typed-racket/base-env/prims-contract.rkt]: I think the trampoline might be doing the wrong thing. Maybe @takikawa has some idea here?

Comment (AlexKnauth) [inline typed-racket-lib/typed-racket/base-env/prims-contract.rkt]: Should that be a different issue? To me it seems better to just have the `Any` fallback.

Comment (samth) [inline typed-racket-lib/typed-racket/base-env/prims-contract.rkt]: I don't think the `Any` fallback makes sense. This seems to be just a bug, which we shouldn't just paper over.

Comment (AlexKnauth) [inline typed-racket-lib/typed-racket/base-env/prims-contract.rkt]: It would be if it generated the wrong contract, yes. But it *is* generating the correct contract in the end.

Comment (samth) [inline typed-racket-lib/typed-racket/base-env/prims-contract.rkt]: No, it's a bug regardless of what's happening in the end. The contract shouldn't be called there.

Comment (AlexKnauth) [inline typed-racket-lib/typed-racket/base-env/prims-contract.rkt]: Ok. But it's still a much smaller bug than the one this fixes.

Comment (samth) [inline typed-racket-lib/typed-racket/base-env/prims-contract.rkt]: Definitely. Hopefully @takikawa will have a suggestion soon.

Comment (AlexKnauth) [inline typed-racket-lib/typed-racket/base-env/prims-contract.rkt]: I was adding some `printf`s and looking at it with @takikawa, we discovered that it work if we do `(identity (cast 3 Integer))`, `(define x (cast 3 Integer))` or any other nested expression context. Also for some reason the repl is broken now. It might be because the lifts create begins, and the trampoline does the contract-generation pass on the first expression of the begin (the lifted contract) before it does anything on the next expressions.

Comment (samth) [inline typed-racket-lib/typed-racket/base-env/prims-contract.rkt]: Is the REPL broken currently, or with this PR?

Comment (AlexKnauth) [inline typed-racket-lib/typed-racket/base-env/prims-contract.rkt]: It's broken with the versions where the thunk errors when it's called to early.

Comment (AlexKnauth) [inline typed-racket-lib/typed-racket/base-env/prims-contract.rkt]: I've patched it up so that the repl works, sandboxed evaluation works, and the docs build now. What I did was change `cast` so that in a top-level context it puts `Any` in the cast-table to start with. But for casts that aren't in a top-level-context, it will be an internal type-checker error for the thunk to be called too early, but that doesn't cause any failures because for non-top-level casts the contract-generation happens at the correct time.

Comment (AlexKnauth) [inline typed-racket-lib/typed-racket/base-env/prims-contract.rkt]: This was just fixed by @takikawa so that the `Any` isn't necessary after all, even at the top level. What he did was have `cast` just expand to `core-cast` within an `#%expression` form, so that `core-cast` will never actually be at the top level, even if `cast` is.

Comment (samth) [inline typed-racket-lib/typed-racket/base-env/prims-contract.rkt]: So now we can just have it error when called too early?

File: typed-racket-lib/typed-racket/base-env/prims-contract.rkt [hunk 9]
```diff
@@ -250,9 +260,27 @@
...
+   ;; By then `tc/#%expression` will have recognized the `casted-expr` property and
+   ;; will have added the casted expression's original type to the cast-table, so
+   ;; that `(cast-table-ref id)` can get that type here.
+   (λ ()
+     (define type-stx
+       (or (cast-table-ref id)
+           (int-err (string-append
+                     "contract-def-property: thunk called too early\n"
+                     "  This should only be called after the type-checking pass has finished."))))
+     `#s(contract-def ,type-stx ,flat? ,maker? typed))))
```
Comment (AlexKnauth) [inline typed-racket-lib/typed-racket/base-env/prims-contract.rkt]: Re: https://github.com/racket/typed-racket/pull/365/files/357e4a12419fbc6af056a77710823bfc53b1a75e#r65783232

> So now we can just have it error when called too early?

Yes, and that's what it does now.

Comment (AlexKnauth) [inline typed-racket-lib/typed-racket/base-env/prims-contract.rkt]: It errors when it's called too early, and it never tries an `Any` fallback in any situation.

**Comment by AlexKnauth** [comment]
I've just squashed all of my commits into one, which got rid of the reverts of reverts.

**Comment by samth** [comment]
This looks great. Ready to merge!

**Comment by AlexKnauth** [comment]
Should I merge it then?
